# Name: Muhammad Hamid Khan
# Roll No: 24i-7352
# Subject: Programing for Business Application
# Course: Master in Business Analytics
# Topic: Python Fundamentals

# ---------------------------------------------------------------------------------------

# 1: Variables in Programming

### 
Data consists of raw facts about something. It could be any information or knowledge that describes something and is referred to as data. To store this information, it needs to be placed in a specific location with a designated name for later use. This is where variables come into play. Variables allow us to assign information a specific name (naming convention) so that it can be accessed later

#### Variables are temporary storage where values are stored. The semantic meaning of variable is varying or changing. We can assign this variable to another variable and can change the value of that variable as well. 

# Example 01: 

In [16]:
my_name = "Muhammad Hamid Khan"

In [18]:
my_name

'Muhammad Hamid Khan'

##### Here I have stored my name in the "my_name" variable. Wherever, and whenever, I want to call it, or use it, I will call my_name, and it automatically gives my name, Because my name is assigned to this variable. 


# Example 02:

In [23]:
my_current_status = "I am currently doing my Master in Business Analytics. Also, I am doing job as Tech Fellow based in Islamabad" 

In [25]:
my_current_status

'I am currently doing my Master in Business Analytics. Also, I am doing job as Tech Fellow based in Islamabad'

##### Here, I have stored piece on information about my profession and present time social status. I will simply call the variable named "my_current_status" when I need to refer or tell about my self to someone. 

# Example 03:

In [32]:
a, b = 20, 30

##### Here, I have assigned 20 and 30 values to two variables a and b respectively.

### Performing some basic arithmatic operations by using a and b.

In [ ]:
# Addition

In [37]:
a + b 

50

In [ ]:
# Subtraction 

In [39]:
a - b

-10

In [ ]:
# Multiplication

a * b

In [ ]:
# Division

In [43]:
a / b

0.6666666666666666

#### Instead of values, used variables which have these values. 

# Decision Making Statements

### Decision-making statements are conditional statements. It is basically, making a decision based on condition and you have options to choose from. It follows if-else, if-elif, and also nested-if. Decision-making statements are very crucial in real-world programming applications. 

# Example 03: 

In [381]:
# Define the correct password
Password = "khan@121314"

# Prompt the user to enter their password
password = str(input("𝙀𝒏𝙩𝒆𝙧 𝙮𝒐𝙪𝒓 𝑷𝙖𝒔𝙨𝒘𝙤𝒓𝙙: "))

# Check if the entered password matches the stored password
if password == Password:
    print("Yᴏᴜ ᴀʀᴇ ᴡᴇʟᴄᴏᴍᴇ ᴛᴏ ᴛʜᴇ Pᴀɢᴇ")
else:
    print("𝑌o𝑢r p𝑎s𝑠w𝑜r𝑑 𝑖n I𝑛c𝑜r𝑟e𝑐t. 𝑈s𝑒 𝑦o𝑢r c𝑜r𝑟e𝑐t C𝑟e𝑑e𝑛t𝑖a𝑙s")


𝙀𝒏𝙩𝒆𝙧 𝙮𝒐𝙪𝒓 𝑷𝙖𝒔𝙨𝒘𝙤𝒓𝙙:  khan@121314


Yᴏᴜ ᴀʀᴇ ᴡᴇʟᴄᴏᴍᴇ ᴛᴏ ᴛʜᴇ Pᴀɢᴇ


# Example 02:

### Suppose You have been assigned to design a program "Assign Labels to Students Based on there Marks".Student will come, give input to the program, they will be labeled based on the conditions set in the program. 

In [80]:
Your_marks = int(input("𝘼𝙙𝙙 𝙮𝙤𝙪𝙧 𝙢𝙖𝙧𝙠𝙨:  "))
if Your_marks >= 90:
    print("𝓔𝔁𝓬𝓮𝓵𝓵𝓮𝓷𝓽")
elif Your_marks < 90  and Your_marks >= 80:
    print("𝒱𝓮𝓇𝔂 𝓖ℴ𝓸𝒹")
elif  Your_marks < 80 and  Your_marks >= 70:
    print("Ɠơơɗ")
elif  Your_marks < 70 and  Your_marks >= 60:
    print("Ƒᶏᶖᶉ")
elif  Your_marks < 60 and  Your_marks >= 50:
    print("𝓟𝒶𝓼𝓈")
else: 
    print("𝔽𝕒𝕚𝕝! 𝕋𝕣𝕪 𝕟𝕖𝕩𝕥 𝕥𝕚𝕞𝕖") 

𝘼𝙙𝙙 𝙮𝙤𝙪𝙧 𝙢𝙖𝙧𝙠𝙨:   45


𝔽𝕒𝕚𝕝! 𝕋𝕣𝕪 𝕟𝕖𝕩𝕥 𝕥𝕚𝕞𝕖


# Example 03:

### Suppose your Finance Manager has assigned you a task to design a program. He wants to give discounts to customers based on some specific conditions. Users have to put the total amount value.
### Conditions:
#### 1: 15% discount If purchases above $5000

#### 2: 10% discount If purchases above $4000

#### 3: 5% discount If purchases above $3000

In [99]:
Total_Amount = int(input("P𝑢t T𝑜t𝑎l A𝑚o𝑢n𝑡 𝑜f y𝑜u𝑟 𝑃u𝑟cℎa𝑠e𝑑 :"))
if Total_Amount >= 5000:
    print("𝚈𝚘𝚞 𝚑𝚊𝚟𝚎 𝚎𝚊𝚛𝚗𝚎𝚍 𝟷𝟻% 𝚘𝚏 𝙳𝚒𝚜𝚌𝚘𝚞𝚗𝚝")
elif Total_Amount >= 4000:
    print("Y𝚈𝚘𝚞 𝚑𝚊𝚟𝚎 𝚎𝚊𝚛𝚗𝚎𝚍 𝟷0% 𝚘𝚏 𝙳𝚒𝚜𝚌𝚘𝚞𝚗𝚝")
elif Total_Amount >= 3000:
    print("𝚈𝚘𝚞 𝚑𝚊𝚟𝚎 𝚎𝚊𝚛𝚗𝚎𝚍 𝟻% 𝚘𝚏 𝙳𝚒𝚜𝚌𝚘𝚞𝚗𝚝")
else:
    print("𝓢𝓸𝓻𝓻𝔂! 𝓨𝓸𝓾 𝓭𝓸 𝓷𝓸𝓽 𝓱𝓪𝓿𝓮 𝓪𝓷𝔂 𝓓𝓲𝓼𝓬𝓸𝓾𝓷𝓽")
    

P𝑢t T𝑜t𝑎l A𝑚o𝑢n𝑡 𝑜f y𝑜u𝑟 𝑃u𝑟cℎa𝑠e𝑑 : 778


𝓢𝓸𝓻𝓻𝔂! 𝓨𝓸𝓾 𝓭𝓸 𝓷𝓸𝓽 𝓱𝓪𝓿𝓮 𝓪𝓷𝔂 𝓓𝓲𝓼𝓬𝓸𝓾𝓷𝓽


# Looping Statements

### Looping statements are used to perform repeated fucntions or actions again and again until the desire/end goal is achieved. Programmers decide the criteria, it works automatically. Looping also worked based on conditions. 
### There are two types of looping (1): For Loop  (2): While Loop

# For Loop
### For Loop neccesserly do not take a condition.  

In [200]:
Product_Price = 35
Product_quantity = pd.Series([23,22,21,38,34,25,22,66,33,13,42,42,12,17,19,21,21])

In [204]:
for Total_Amount in Product_quantity:
    Total_Amount = Product_Price * Product_quantity

In [216]:
Total_Amount = pd.DataFrame(Total_Amount)

In [218]:
Total_Amount.columns = ["Total_Amount"]

In [220]:
Total_Amount

Total_Amount
0            805
1            770
2            735
3           1330
4           1190
5            875
6            770
7           2310
8           1155
9            455
10          1470
11          1470
12           420
13           595
14           665
15           735
16           735

# While Loop

#### Suppose you have been assigned to automate the "Charity Collection Pool" for your office. Your task is to apply the check, not to add charity to the pool once it reaches the amount of $100.

#### While loop neccesserly takes the condition. But, it usually requires initialization. 

In [296]:
Amount = 0
while Amount < 100:
    Amount_Added = input("𝙋𝒍𝙚𝒂𝙨𝒆 𝑬𝙣𝒕𝙚𝒓 𝒀𝙤𝒖𝙧 𝘿𝒐𝙣𝒂𝙩𝒊𝙤𝒏 𝑨𝙢𝒐𝙪𝒏𝙩: ")
    Amount_Added = int(Amount_Added)  
    Amount += Amount_Added 

print(f"𝓦𝓮 𝓗𝓪𝓿𝓮 𝓒𝓸𝓵𝓵𝓮𝓬𝓽𝓮𝓭 𝓽𝓱𝓮 𝓓𝓮𝓼𝓲𝓻𝓮𝓭 𝓕𝓾𝓷𝓭𝓼","$",Amount)


𝙋𝒍𝙚𝒂𝙨𝒆 𝑬𝙣𝒕𝙚𝒓 𝒀𝙤𝒖𝙧 𝘿𝒐𝙣𝒂𝙩𝒊𝙤𝒏 𝑨𝙢𝒐𝙪𝒏𝙩:  44
𝙋𝒍𝙚𝒂𝙨𝒆 𝑬𝙣𝒕𝙚𝒓 𝒀𝙤𝒖𝙧 𝘿𝒐𝙣𝒂𝙩𝒊𝙤𝒏 𝑨𝙢𝒐𝙪𝒏𝙩:  43
𝙋𝒍𝙚𝒂𝙨𝒆 𝑬𝙣𝒕𝙚𝒓 𝒀𝙤𝒖𝙧 𝘿𝒐𝙣𝒂𝙩𝒊𝙤𝒏 𝑨𝙢𝒐𝙪𝒏𝙩:  23


𝓦𝓮 𝓗𝓪𝓿𝓮 𝓒𝓸𝓵𝓵𝓮𝓬𝓽𝓮𝓭 𝓽𝓱𝓮 𝓓𝓮𝓼𝓲𝓻𝓮𝓭 𝓕𝓾𝓷𝓭𝓼 $ 110


# Functions in Programming

# Example 01:

#### Suppose you have assigned to design a function to build invoice based on list of items price and a tax rate. 

In [350]:
def Invoice(Item_Price):
    if Item_Price >= 500:
        total_price = Item_Price * 1.15  # 15% tax
        print(f"𝙏𝒐𝙩𝒂𝙡 𝙋𝒓𝙞𝒄𝙚 𝙬𝒊𝙩𝒉 𝑻𝙖𝒙: {total_price:.2f}")
    elif Item_Price >= 400:
        total_price = Item_Price * 1.10  # 10% tax
        print(f"𝗧𝗼𝘁𝗮𝗹 𝗣𝗿𝗶𝗰𝗲 𝘄𝗶𝘁𝗵 𝗧𝗮𝘅: {total_price:.2f}")
    elif Item_Price >= 300:
        total_price = Item_Price * 1.05  # 5% tax
        print(f"𝗧𝗼𝘁𝗮𝗹 𝗽𝗿𝗶𝗰𝗲 𝘄𝗶𝘁𝗵 𝗧𝗮𝘅: {total_price:.2f}")
    else: 
        print("G𝑜 𝑇o F𝑖n𝑎n𝑐e M𝑎n𝑎g𝑒r o𝑛 𝐷e𝑠k N𝑜# 01")


In [352]:

Purchased_Item_Price = int(input("𝓐𝓭𝓭 𝓘𝓽𝓮𝓶 𝓟𝓻𝓲𝓬𝓮: "))
Invoice(Purchased_Item_Price)


𝓐𝓭𝓭 𝓘𝓽𝓮𝓶 𝓟𝓻𝓲𝓬𝓮:  334


𝗧𝗼𝘁𝗮𝗹 𝗽𝗿𝗶𝗰𝗲 𝘄𝗶𝘁𝗵 𝗧𝗮𝘅: 350.70


#### Above code fixed the Tax Amount based on Item Price. It basically designed to apply high tax rates on expensive items and vice versa. 

# Example 02: 

### Suppose your Python Teacher has asked you to design a Program (Function)  which takes marks of subjects and add them. This function must give marks as a percentage of the total. 

In [53]:
def marks(Subject1, Subject2, Subject3, Subject4):
    percentage = ((Subject1 + Subject2 + Subject3 + Subject4) / 400) * 100
    print(percentage)

In [61]:
Marks1 = int(input("Add Marks for subject 01 : "))
Marks2 = int(input("Add Marks for subject 02 : "))
Marks3 = int(input("Add Marks for subject 03 : "))
Marks4 = int(input("Add Marks for subject 04 : "))
marks(Marks1, Marks2, Marks3, Marks4)

Add Marks for subject 01 :  80
Add Marks for subject 02 :  79
Add Marks for subject 03 :  78
Add Marks for subject 04 :  77


78.5


# Object Oriented Programming Concepts

#### Object-oriented programming is a technique that behaves like a real-world object. In Python, OOP provides an effective and efficient template for complex programming and coding efforts. Classes are a crucial component of OOP. Classes have behaviors and properties that make them identical in programming. In addition, objects are part of classes, where each object is unique but holds the same properties and behaviors as its parent class.

#### For example Pandas, Numpy, Matplotlib, and Seaborn libraries in Python.

# Algorithmic Approach to Solve a Problem

#### In Python functions and programs are designed by following the step by step approach. These are aimed to provide effective and efficient solutions to real world problems. This approach includes essential components and also requires neccessary arrangements. 

#### (Input)-----------------------[Steps to followed]-----------------------(Output)
#### For example: You are assigned with a task to develop a program which will detect the item as defective or non_defective. The Detector, basically an algorithum at the backed has been properly trained. That is known as Supervised Machine Learning.   

# Introduction to Python

#### Python is a programming language. It is open-source, and also oop language. It has large number of standard and comprehensive libraries. In addition, it has IDE's including Pycharm and Anaconda Navigator. Anaconda Navigator is basically web base distributor. It provides different comprehensive dependencies. 

# Intro to Jupyter Notebook

#### Jupyter Notebook is browser base interactive python interprator. It has very effective and efficient user interface. It has comprehensive build-in dependencies which allow python developer to program complex problem very effectively. It supports the executive programing. 

# Variables and Data-types in Python

#### Variables are temporary storage where values are stored. The semantic meaning of variable is varying or changing. We can assign this variable to another variable and can change the value of that variable as well.

In [88]:
Python_Sir = ("Mr.Faheem-ur-Rehman")

In [90]:
print(f"{Python_Sir} is highly recommended instructor for Python Language.")

Mr.Faheem-ur-Rehman is highly recommended instructor for Python Language.


In [94]:
Python_Sir = ("Muhammad Hamid Khan Niazi")
print(f"{Python_Sir} is highly recommended instructor for Python Language.")

Muhammad Hamid Khan Niazi is highly recommended instructor for Python Language.


#### So, here in the above two code snippets, previously "Python_Sir" variable had the value "Mr.Faheem-ur-Rehman". But later, it got changed to "Muhammad Hamid Khan Niazi" via variable reassignment. 

In [98]:
my_name = call_me = ("Muhammad Hamid Khan Niazi")

In [100]:
my_name

'Muhammad Hamid Khan Niazi'

In [102]:
call_me

'Muhammad Hamid Khan Niazi'

#### We can assign value to two variables in same line of code, that is called as Chained Assignment. That is why, Python is called as dynamic programming language. 

## Data_Types

#### In Python every data has datatype. These datatypes make variables identical from other variables. Based on these datatypes, different complex programming techniques and libraries are also applied. 

#### In Python there are numerious datatypes. But few are in common use, including, string(str), interger(int), boollen, and float. 

### Integer

##### Any discrete numeric values ranging from positive infinity to negative infinity. For example 2323,24234,6463,124353,34534. In addition, the "int" method is used to specify anything as an "Integer" in a Python programming language. 

In [144]:
Hostel_Charges = 25000 

In [162]:
print(type(Hostel_Charges))


<class 'int'>


### Float

##### Floats are decimal numeric values. These are points that are also useful while dealing with values, and algorithms dealing with close values which are mainly in decimals. 
##### For example: 24.33, 35.11, 2.34, and many more. Student CGPA is a very glaring example of floating point numbers.

In [166]:
Bs_CGPA = 3.62

In [168]:
print(type(Bs_CGPA))

<class 'float'>


### Boolean

#### Boolean consists of two values which are "True" and "False". These are logical values that are mainly used in conditional statements. Commonly, "1" is assigned to "Ture" and "0" is assigned to "False".

In [133]:
a = 5 
b = 10

In [135]:
a > b

False

In [128]:
a < b

True

### String 

##### String consists of characters that are enclosed in double("  ") or single quotes ('  '). The string is mainly used for alphabetical characters. It uses the "str" method to specify something as a string. 

In [142]:
my_name = "Muhammad Hamid Khan"
my_name

'Muhammad Hamid Khan'

In [170]:
print(type(my_name))

<class 'str'>


# Operators in Python

### Operators are build_in functions in Python that perform specific operations. Commonly, we have three operators in Python, named Arithmetic Operators, Relational Operators, and Logical Operators. 

## Arithmetic Operators
#### Basically it performs mathematical operations like + , -, * and / on the data. 

In [178]:
a = 20
b = 10

In [180]:
# Addition
a+b

30

In [182]:
# Subtraction
a-b

10

In [184]:
# Multiplication
a*b

200

In [186]:
# Division
a / b

2.0

## Relational Operators
#### Relational operators play very key role in conditional statements. It includes, >, <, <+, >=, != and ==. The following examples give the output as Ture or False based on the condition understudy. 

In [193]:
c = 22
d = 31

In [195]:
# Greater Then
c > d

False

In [197]:
# Less Then
c < d

True

In [201]:
# Greater then or equal to
c <= d


True

In [204]:
# Less then or equal to
c >= d

False

In [206]:
# Not equal to 
c != d

True

In [208]:
# Equal To
c == d

False

# Logical Operators 

### Logical operators are somehow also used for conditions. These include, "And" (&) which is basically multiply (*), and "Or" which is (+). In addition, And Operator gives the "True Results" if both the conditions are true otherwise "False". And, Or Operator gives "True Results" if any of the one conditions is true. In Or if both are "False" then it gives the "False Results".   

## Example 01: 

## AND
### Suppose, two sisters will receive 50% of the fee waiver if they both get 80% marks in sessional exams.  

In [242]:
sister_A_marks = int(input("Enter your marks"))
sister_B_marks = int(input("Enter your marks"))

if sister_A_marks >= 80 and sister_B_marks >= 80:
    print("𝘾𝙤𝙣𝙜𝙧𝙖𝙩𝙪𝙡𝙖𝙩𝙞𝙤𝙣𝙨! 𝙔𝙤𝙪 𝙝𝙖𝙫𝙚 𝙜𝙤𝙩 50% 𝙤𝙛 𝙩𝙝𝙚 𝙛𝙚𝙚 𝙬𝙖𝙫𝙞𝙚𝙧")
else:
    print("𝓢𝓸𝓻𝓻𝔂 𝓨𝓸𝓾 𝓪𝓻𝓮 𝓷𝓸𝓽 𝓔𝓵𝓲𝓰𝓲𝓫𝓵𝓮")

Enter your marks 84
Enter your marks 79


𝓢𝓸𝓻𝓻𝔂 𝓨𝓸𝓾 𝓪𝓻𝓮 𝓷𝓸𝓽 𝓔𝓵𝓲𝓰𝓲𝓫𝓵𝓮


## OR
### Suppose, A student can get a 40% fee waiver if he is from a rural area or has marks greater than 70%. 

In [279]:
Marks = int(input("Enter your marks: "))
Residential_Area = str(input("Enter your Residential Area (Rural or Urban): ")).strip().lower()

if Marks >= 70 or Residential_Area == "rural":
    print("𝘾𝙤𝙣𝙜𝙧𝙖𝙩𝙪𝙡𝙖𝙩𝙞𝙤𝙣𝙨! 𝙔𝙤𝙪 𝙝𝙖𝙫𝙚 𝙜𝙤𝙩 70% 𝙤𝙛 𝙩𝙝𝙚 𝙛𝙚𝙚 𝙬𝙖𝙫𝙞𝙚𝙧")
else:
    print("𝓢𝓸𝓻𝓻𝔶 𝓨𝓸𝓾 𝓪𝓻𝓮 𝓷𝓸𝓽 𝓔𝓵𝓲𝓰𝓲𝓫𝓵𝓮")


Enter your marks:  56
Enter your Residential Area (Rural or Urban):  rural


𝘾𝙤𝙣𝙜𝙧𝙖𝙩𝙪𝙡𝙖𝙩𝙞𝙤𝙣𝙨! 𝙔𝙤𝙪 𝙝𝙖𝙫𝙚 𝙜𝙤𝙩 70% 𝙤𝙛 𝙩𝙝𝙚 𝙛𝙚𝙚 𝙬𝙖𝙫𝙞𝙚𝙧


# Tokens in Python

### Tokens are meaningful laterials in Python that are self identical. Following are the few tokens.
#### 01: Keywords
#### 02: Identifiers
#### 03: Literals
#### 04: Operators 

##  Keywords

#### Keywords are special reserved words in Python. These reserved words are treated special having identical meanings. Commonly, these are variables or naming conventions that have been assigned to these words by the developers. That is why, these are called as Keywords. So, we cannot again assign the same names to the variables. Otherwise, it will give typo or syntax error.  
#### Following are the few keywords. 
#### False, True, And, Or, If, else, def, while, for and many more. Keywords are highlighted "Green" in Python Jupyter Notebook. 

In [ ]:
if 

## See it is in Green Color. 

In [289]:
if = 20

SyntaxError: invalid syntax (2644998005.py, line 1)

In [291]:
True = 33

SyntaxError: cannot assign to True (2174846472.py, line 1)

In [ ]:
## Look at code lines 289 and 291, these gave the syntax error, because "if" is keyword. We cannot take it as variable. 

In [297]:
name = def = "Muhammad Hamid Khan"

SyntaxError: invalid syntax (1296032336.py, line 1)

In [ ]:
## Look, it also gave error with Chained Variable. Because "def" is keuwords so it can't be used as variable. 

##  Identifiers

#### Identifiers are naming conventions in Python. These names are assigned to functions, variables, and objects based on conditions. 
#### We have some rules for naming variables. 1: Can't start with numbers 2: Can't start with special characters 3: Can't have special character in-between 4: Onlt starts with alphabetic character 5: Start with Under_Score 6: Only Under_score is allowed to use for example my_name. In addition, these identifiers are case-sensitive. 

In [ ]:
# 1: Can't start with numbers 

In [308]:
4_my_name = 'Hami Niaxi'
4_my_name

SyntaxError: invalid decimal literal (3746239209.py, line 1)

In [ ]:
# 2: Can't start with special characters 

In [314]:
@my_name = "Hami Niaxi Watta Khal"
@my_name

SyntaxError: invalid syntax. Maybe you meant '==' or ':=' instead of '='? (592541476.py, line 1)

In [ ]:
# 3: Can't have special character in-between

In [316]:
my_@_name = "Hamid Niazi"
my_@_name

SyntaxError: cannot assign to expression here. Maybe you meant '==' instead of '='? (1232209899.py, line 1)

In [ ]:
# 4: Onlt starts with alphabetic character 

In [318]:
myname = "Muhammad Hamid Khan Niazi Watta Khal all the way from Mianwali"
myname 

'Muhammad Hamid Khan Niazi Watta Khal all the way from Mianwali'

In [ ]:
# 5: Start with Under_Score

In [320]:
_myname = "Hamid Niazi"
_myname

'Hamid Niazi'

In [ ]:
# 6: Only Under_score is allowed in-between alphabatic characters 

In [340]:
my_name = """ Muhammad Hamid Khan Niazi Watta Khal from Mianwali. He is currently doing his master's in Business Analytics. His research interests are
Machine Learning and Computer Vision """
my_name

" Muhammad Hamid Khan Niazi Watta Khal from Mianwali. He is currently doing his master's in Business Analytics. His research interests are\nMachine Learning and Computer Vision "

## Literals

#### Python literals are basically constants, that are static in nature. Let's suppose we have numeric, string, and boolean literals.  Simply, the values that are stored in variables are called literals. 
#### For example; my_name = "Muhammad Hamid Khan". Here, my_name is variable and value Muhammad Hamid Khan is literal.


In [346]:
Great_Leader = "Imran Ahmed Khan Niazi "
Great_Leader

'Imran Ahmed Khan Niazi'

In [ ]:
## Here 'Great_Leader' is variable, and 'Imran Ahmed Khan Niazi' is a literial. (String Literal)

In [348]:
my_marks = 78
my_marks

78

In [ ]:
## Here 'my_marks' is variable, and '78' is a literial. (Numeric Literal)

# Strings in Python

#### Strings are collections of characters that are in sequence. These characters are enclosed in single, double, and triple quotes ' ', " ", and "'  "' respectively. 

#### Indexing in string (Indexing starts with 0)

#### Slicing in string also starts with 0 but it takes range(0:10)
#### Indexing and slicing is basically accessing a value from a variable at any position. 


In [356]:
## Single quote
call_me1 = 'Hami Niaxi'
call_me1

'Hami Niaxi'

In [358]:
## Double quotes
call_me2 = "Hamid Niaxi"
call_me2

'Hamid Niaxi'

In [360]:
## Triple quotes
call_me3 = "Hamid Niazi"
call_me3

'Hamid Niazi'

In [367]:
## Multi Lines
call_me_multi = """ My Name is Muhammad Hamid Khan Niazi
I am a professional Business Analyst. 
My research interests are Machine Learning and Computer Vision
"""
call_me_multi

' My Name is Muhammad Hamid Khan Niazi\nI am a professional Business Analyst. \nMy research interests are Machine Learning and Computer Vision\n'

#### Indexing 

In [377]:
call_me_indx= ["Muhammad", "Hamid", "Khan", "Niazi"]
call_me_indx

['Muhammad', 'Hamid', 'Khan', 'Niazi']

In [383]:
call_me_indx[0]

'Muhammad'

In [385]:
call_me_indx[1]

'Hamid'

In [387]:
call_me_indx[2]

'Khan'

In [389]:
call_me_indx[3]

'Niazi'

In [391]:
call_me_indx[0], call_me_indx[1], call_me_indx[2], call_me_indx[3]

('Muhammad', 'Hamid', 'Khan', 'Niazi')

In [393]:
call_me_indx[3], call_me_indx[2], call_me_indx[1], call_me_indx[0]

('Niazi', 'Khan', 'Hamid', 'Muhammad')

In [395]:
call_me_indx[2],call_me_indx[0], call_me_indx[1], call_me_indx[2], call_me_indx[3]

('Khan', 'Muhammad', 'Hamid', 'Khan', 'Niazi')

#### Slicing 

In [397]:
call_me_slic = """ My Name is Muhammad Hamid Khan Niazi
I am a professional Business Analyst. 
My research interests are Machine Learning and Computer Vision
"""
call_me_slic

' My Name is Muhammad Hamid Khan Niazi\nI am a professional Business Analyst. \nMy research interests are Machine Learning and Computer Vision\n'

In [421]:
call_me_slic[:37]

' My Name is Muhammad Hamid Khan Niazi'

In [417]:
call_me_slic[38:75]

'I am a professional Business Analyst.'

In [419]:
call_me_slic[76:]

'\nMy research interests are Machine Learning and Computer Vision\n'

### Length of the String 
#### len() method is used to count the numbers of character in Python. 

In [424]:
call_me_len = """ My Name is Muhammad Hamid Khan Niazi
I am a professional Business Analyst. 
My research interests are Machine Learning and Computer Vision
"""
call_me_len

' My Name is Muhammad Hamid Khan Niazi\nI am a professional Business Analyst. \nMy research interests are Machine Learning and Computer Vision\n'

In [429]:
len(call_me_len)

140

In [ ]:
###  Variable call_me_len has 140 characters 

### Converting string characters into lowercase.
#### lower() method converts the character into lowercase. For example; lower('Muhammad Hamid Khan') ..'muhammad hamid khan'

In [433]:
making_lower = 'MUHAMMAD HAMID KHAN'
making_lower

'MUHAMMAD HAMID KHAN'

In [437]:
(making_lower).lower()

'muhammad hamid khan'

### Converting string characters into Uppercase.
#### upper() method converts the character into uppercase. For example; upper('muhammad hamid khan') ..'MUHAMMAD HAMID KHAN'

In [441]:
making_upper = 'muhammad hamid khan'
making_upper 

'muhammad hamid khan'

In [443]:
making_upper.upper()

'MUHAMMAD HAMID KHAN'

### Replacing a Sub-string
#### .replace() method takes two parameters such as New_Value, and Old_Value. It is used to make changes or corrections in the existing strings. 

In [448]:
making_replacement = "Muhammad Hami Niaxi"
making_replacement 

'Muhammad Hami Niaxi'

In [468]:
making_replacement.replace("Hami Niaxi", "Hamid Niazi")

'Muhammad Hamid Niazi'

### Number of occurance of Sub-string 
#### .count() method is used to count the accurance of any character in the string. It simply count which character has been accured how many times. 

In [475]:
my_intro = "My name is Muhammad Hamid Khan Niazi. I belonged to a Niazi tribe. Niazi tribe migrated from Afghanistan in the early 18th Century."
my_intro.count("Niazi")

3

### Findind a index of the String
#### .find() method in used to get the index of the string. It basically tells about the exact position of that word or character in the substring. 

In [490]:
known_as  = 'Muhammad Hamid Khan Niazi'
known_as.find('Muhammad')

0

In [492]:
known_as.find('Hamid')


9

In [494]:
known_as.find('Khan')

15

In [496]:
known_as.find('Niazi')

20

In [ ]:
# Above outputs are basically telling about the starting postion of that words or character in the sub-string. 

### Splitting the string
#### .split() method splits the string into different parts. It basically splits the string into sub-strings.

In [513]:
ismy_gramii  = 'Muhammad, Hamid ,Khan ,Niazi'
splitted_name = ismy_gramii.split(",")
splitted_name


['Muhammad', ' Hamid ', 'Khan ', 'Niazi']

In [ ]:
## Indexing can varify the splitting 

In [517]:
ismy_gramii[2]  

'h'

In [515]:
splitted_name[2]

'Khan '

In [ ]:
## The outputs of the indexing validate the successful implementation of splitting method. 

# Python Tuples

#### Python tuple is a data structure in Python. It is an ordered collection of heterogeneous values that are stored in a single variable and enclosed in round brackets *( )*. In addition, tuples are immutable which means that once the values are stored they can't be changed or altered.  

In [1]:
tuple_01 = (22486, 'Muhammad Hamid Khan', 'MS-7352', 3466993886, 23, 'Unmarried')
tuple_01

(22486, 'Muhammad Hamid Khan', 'MS-7352', 3466993886, 23, 'Unmarried')

In [35]:
type(tuple_01)

tuple

#### Indexing
#### Indexing is basically accessing any value from tuple at any position.It takes the numeric address of the value like [23].

In [45]:
tuple_01[2]

'MS-7352'

In [47]:
tuple_01[5]

'Unmarried'

In [5]:
tuple_01[3]

3466993886

#### Slicing 
#### Slicing is also used to access any value from the tuple. But it takes the address in a range. Like [2:17], it will start from a character on numeric address 3 and will access up to 17. But it does not include the character at numeric address 17.

In [11]:
tuple_2 = ("My name is Muhammad Hamid Khan Niazi. I belonged to a Niazi tribe. Niazi tribe migrated from Afghanistan in the early 18th Century.")
tuple_2

'My name is Muhammad Hamid Khan Niazi. I belonged to a Niazi tribe. Niazi tribe migrated from Afghanistan in the early 18th Century.'

In [64]:
tuple_2[0:]

'My name is Muhammad Hamid Khan Niazi. I belonged to a Niazi tribe. Niazi tribe migrated from Afghanistan in the early 18th Century.'

In [66]:
tuple_2[:-1]

'My name is Muhammad Hamid Khan Niazi. I belonged to a Niazi tribe. Niazi tribe migrated from Afghanistan in the early 18th Century'

In [70]:
tuple_2[0:31]

'My name is Muhammad Hamid Khan '

In [80]:
tuple_2[38:66] 

'I belonged to a Niazi tribe.'

####  Tuple Modifying

In [27]:
tuple_2[19:25]="Hami"

TypeError: 'str' object does not support item assignment

#### This showed this error becuase tuples are imputables. 

### Finding lenght of tuple 

In [35]:
tuple_2 

'My name is Muhammad Hamid Khan Niazi. I belonged to a Niazi tribe. Niazi tribe migrated from Afghanistan in the early 18th Century.'

In [37]:
len(tuple_2)

131

##### Total there are 131 characters are in tuple_2.

#### Concatenating Tuple

In [74]:
tuple_a = (" Muhammad Hamid")
tuple_b = (" Niazi")
    

In [76]:
tuple_can = tuple_a + tuple_b
tuple_can

' Muhammad Hamid Niazi'

#### Tuple cancatenating is possible which is practiced above in abovr line. Elements from left side of the + operator come first and otherwise.  

### Repeating Tuple element in the Tuple 

In [78]:
tuple_can

' Muhammad Hamid Niazi'

In [80]:
tuple_can * 3

' Muhammad Hamid Niazi Muhammad Hamid Niazi Muhammad Hamid Niazi'

### Repeating and cancatenating in the same line of code

In [72]:
tuple_a1 = "Imran Khan Niazi "
tuple_a2 = " 804"
tuple_a1*2 + tuple_a2

'Imran Khan Niazi Imran Khan Niazi  804'

## Tuple Functions
### Minimum Value

In [87]:
tup_min = (34,34,22,13,6,4,2,66,33,24)
min(tup_min)

2

##### Mininmum value is "2".

### Maximum Value

In [91]:
tup_max = (34,34,22,13,6,4,2,66,33,24)
max(tup_max)

66

##### Maximum value is "66"

# List in Python

#### List is the collection of elements that are enclosed in square brackets. [  ]
#### List are mutable. It means taht values inside the list is changable. 

In [99]:
lst_1 = ['Imran', 'Ahmad', 'Khan', 'Niazi', 'Qadii_number', 804]
lst_1

['Imran', 'Ahmad', 'Khan', 'Niazi', 'Qadii_number', 804]

In [101]:
type(lst_1)

list

### Extracting individual elements from the list

In [108]:
# Indexing
lst_1[0]

'Imran'

In [110]:
lst_1[3]

'Niazi'

In [116]:
# Slicing 
lst_1[0:4]

['Imran', 'Ahmad', 'Khan', 'Niazi']

In [120]:
lst_1[4:7]

['Qadii_number', 804]

### Changing the element at index position

In [125]:
lst_1[1]="Hamid"
lst_1

['Imran', 'Hamid', 'Khan', 'Niazi', 'Qadii_number', 804]

### Appending a new element

In [133]:
lst_1.append("A great leader")
lst_1

['Imran', 'Hamid', 'Khan', 'Niazi', 'Qadii_number', 804, 'A great leader']

### Poping the last element

In [138]:
lst_1.pop()
lst_1

['Imran', 'Hamid', 'Khan', 'Niazi', 'Qadii_number']

In [140]:
lst_1.append(804)
lst_1

['Imran', 'Hamid', 'Khan', 'Niazi', 'Qadii_number', 804]

### Reversing element of a list

In [143]:
lst_1.reverse()
lst_1

[804, 'Qadii_number', 'Niazi', 'Khan', 'Hamid', 'Imran']

### Insert element at a specifies index

In [146]:
lst_1

[804, 'Qadii_number', 'Niazi', 'Khan', 'Hamid', 'Imran']

In [148]:
lst_1.insert(6, "from Mianwali")
lst_1

[804, 'Qadii_number', 'Niazi', 'Khan', 'Hamid', 'Imran', 'from Mianwali']

### Sorting a list 

In [161]:
lst_num =["Hamid Niazi", "Imran Niazi", "Sir Faheem shb","Junaid Khalil","Malik", "Ali"] 

In [163]:
lst_num.sort()
lst_num

['Ali',
 'Hamid Niazi',
 'Imran Niazi',
 'Junaid Khalil',
 'Malik',
 'Sir Faheem shb']

### Concatenating lists

In [167]:
lst_can1 = ['Hamid']
lst_can2 = ['Khan']
lst_can3 = ['Niazi']
lst_sum = lst_can1 + lst_can2 + lst_can3
lst_sum

['Hamid', 'Khan', 'Niazi']

### Repeating Elements

In [170]:
lst_sum

['Hamid', 'Khan', 'Niazi']

In [172]:
lst_sum*2

['Hamid', 'Khan', 'Niazi', 'Hamid', 'Khan', 'Niazi']

# Dictionary in Python

#### Dictionary is an un-ordered collection of key-value pairs enclosed wiht {}.
#### Dictionary is mutable.

In [177]:
## Name , age 
dic_1 = {'Hamid': 23, 'Nind': 24, 'Mansoor': 28, 'Muiz': 26}
dic_1

{'Hamid': 23, 'Nind': 24, 'Mansoor': 28, 'Muiz': 26}

### Extracting keys

In [180]:
dic_1.keys()

dict_keys(['Hamid', 'Nind', 'Mansoor', 'Muiz'])

### Extracting values

In [183]:
dic_1.values()

dict_values([23, 24, 28, 26])

### Adding a new element

In [186]:
dic_1['Faisal']=29

In [188]:
dic_1

{'Hamid': 23, 'Nind': 24, 'Mansoor': 28, 'Muiz': 26, 'Faisal': 29}

### Chaning an existing element

In [191]:
dic_1

{'Hamid': 23, 'Nind': 24, 'Mansoor': 28, 'Muiz': 26, 'Faisal': 29}

In [195]:
dic_1['Hamid']=24
dic_1  ## Age changes from 23 to 24

{'Hamid': 24, 'Nind': 24, 'Mansoor': 28, 'Muiz': 26, 'Faisal': 29}

### Updating one dictionary element with other

In [198]:
dic_2={'Haider': 43, 'Nadir': 52, 'Mahnoor': 45}
dic_1.update(dic_2)
dic_1

{'Hamid': 24,
 'Nind': 24,
 'Mansoor': 28,
 'Muiz': 26,
 'Faisal': 29,
 'Haider': 43,
 'Nadir': 52,
 'Mahnoor': 45}

In [200]:
### Popping an element 
dic_1.pop("Nadir")
dic_1

{'Hamid': 24,
 'Nind': 24,
 'Mansoor': 28,
 'Muiz': 26,
 'Faisal': 29,
 'Haider': 43,
 'Mahnoor': 45}

# Set in Python

#### Set in an unordered and unindexed collection of elements enclosed {}
#### Duplicates are not allowed in set

In [206]:
set_1 = { 1,2,1,'Hamid', "Khan", "Hamid"}
set_1

{1, 2, 'Hamid', 'Khan'}

### Adding a new element 

In [209]:
set_1.add('Niazi')
set_1

{1, 2, 'Hamid', 'Khan', 'Niazi'}

### Updating multiple elements

In [216]:
set_1.update(['From', 'Mianwali', 'Living in Islamabad'])
set_1

{' ',
 1,
 2,
 'A',
 'B',
 'F',
 'From',
 'Hamid',
 'I',
 'Khan',
 'L',
 'Living in Islamabad',
 'M',
 'Mianwali',
 'Niazi',
 'S',
 'a',
 'b',
 'd',
 'doing MS BA',
 'g',
 'i',
 'l',
 'm',
 'n',
 'o',
 'r',
 's',
 'v',
 'w'}

### Removing an element 

In [219]:
set_1.remove('Hamid')
set_1

{' ',
 1,
 2,
 'A',
 'B',
 'F',
 'From',
 'I',
 'Khan',
 'L',
 'Living in Islamabad',
 'M',
 'Mianwali',
 'Niazi',
 'S',
 'a',
 'b',
 'd',
 'doing MS BA',
 'g',
 'i',
 'l',
 'm',
 'n',
 'o',
 'r',
 's',
 'v',
 'w'}

### Union of two set

In [222]:
s1 = {1,2,3}
s2 = {4,5,6}
s1.union(s2)

{1, 2, 3, 4, 5, 6}

In [224]:
s3= {1,2,3,5}
s4 = {2,3,4,5,6}
s3.intersection(s4)

{2, 3, 5}

# If Statement in Python

### If condition is applied when we have conditions

#### Simple if 

In [240]:
a = 30
b = 20

In [242]:
if a > b:
    print('a in greater then b')
    

a in greater then b


#### If else

In [247]:
if a < b:
    print('a is greater')
else:
    print('a is lesser')

a is lesser


### Elif (if else, else)

In [250]:
c = 30

In [260]:
if (a > b) and (a > c):
    print("a is greater")
elif (a == b) or (a == c):  
    print('a is equal to b or c')
else:
    print('a is less')

a is equal to b or c


### 1. Using if-else with a Tuple

In [266]:
my_tuple = (10, 20, 30)
if my_tuple[0] > 15:
    print("The first element is greater than 15.")
else:
    print("The first element is not greater than 15.")


The first element is not greater than 15.


### 2. Using if-else with a List

In [268]:
my_list = [5, 15, 25]
if my_list[1] < 20:
    print("The second element is less than 20.")
else:
    print("The second element is not less than 20.")


The second element is less than 20.


### 3. Using if-else with a Dictionary

In [271]:
my_dict = {'a': 100, 'b': 200}
if my_dict['b'] == 200:
    print("The value of key 'b' is 200.")
else:
    print("The value of key 'b' is not 200.")

The value of key 'b' is 200.


# Looping Statements in Python

### Looping statements are used to repeat the cetain tasks multiple times. We have 'For' and 'While' loops

## While Loop

In [278]:
counter = 1
while counter <= 15:
    print(counter) 
    counter += 1    

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15


### While loop for List

In [283]:
my_list = [10, 20, 30, 40, 50]
index = 0
while index < len(my_list):
    print(my_list[index])
    index += 1


10
20
30
40
50


## For Loop

In [285]:
for i in range(1, 6):
    print(i)


1
2
3
4
5


### For Loop for list

In [289]:
fruits = ["apple", "banana", "cherry", "date"]
for fruit in fruits:
    print(fruit)


apple
banana
cherry
date


# Basic Problems in Python

In [292]:
## Checking even add odd

In [294]:
def check_even_odd(number):
    if number % 2 == 0:
        return "Even"
    else:
        return "Odd"

num = int(input("Enter a number: "))
result = check_even_odd(num)
print(f"The number {num} is {result}.")


Enter a number:  23


The number 23 is Odd.


In [296]:
## Checking number +ve, -ve and 0

In [298]:
def check_number(num):
    if num > 0:
        return "Positive"
    elif num < 0:
        return "Negative"
    else:
        return "Zero"


number = float(input("Enter a number: "))
result = check_number(number)
print(f"The number {number} is {result}.")


Enter a number:  4


The number 4.0 is Positive.


In [300]:
## Factorial of number

In [302]:
def factorial(n):
    if n < 0:
        return "Factorial is not defined for negative numbers."
    elif n == 0 or n == 1:
        return 1
    else:
        return n * factorial(n - 1)

number = 5
print(f"The factorial of {number} is {factorial(number)}")


The factorial of 5 is 120


In [304]:
## Check if it plandraine

In [306]:
def is_palindrome(string):
    cleaned_string = ''.join(string.split()).lower()
    return cleaned_string == cleaned_string[::-1]

input_string = input("Enter a string: ")
if is_palindrome(input_string):
    print(f'"{input_string}" is a palindrome.')
else:
    print(f'"{input_string}" is not a palindrome.')


Enter a string:  3


"3" is a palindrome.


# Functions in Python

In [309]:
## 1. Calculate Factorial

In [311]:
def factorial(n):
    if n == 0 or n == 1:
        return 1
    else:
        return n * factorial(n - 1)


num = 5
print(f"The factorial of {num} is {factorial(num)}")


The factorial of 5 is 120


In [313]:
## 2. Check for Prime Number

In [315]:
def is_prime(num):
    if num <= 1:
        return False
    for i in range(2, int(num ** 0.5) + 1):
        if num % i == 0:
            return False
    return True


number = 17
print(f"{number} is a prime number: {is_prime(number)}")


17 is a prime number: True


In [317]:
## 3. Count Vowels in a String

In [319]:
def count_vowels(string):
    vowels = 'aeiouAEIOU'
    count = sum(1 for char in string if char in vowels)
    return count



test_string = "Hello, how many vowels are in this sentence?"
print(f"Number of vowels in the string: {count_vowels(test_string)}")


Number of vowels in the string: 13


# THE END # 